refactor(core): static-query schematic should check templates

devversion opened the pull request:
Queries can technically be also accessed within component templates
e.g.

```html
<my-comp [binding]="myQuery"></my-comp>
```

In that case the query with the property "myQuery" is accessed
statically and needs to be marked with `static: true`. There are
other edge cases that need to be handled as the template property
read doesn't necessarily resolve to the actual query property.

For example:

```html
<foo #myQuery></foo>
<my-comp [binding]="myQuery"></my-comp>
```

In this scenario the binding doesn't refer to the actual query
because the template reference variable takes precedence. The
query doesn't need to be marked with "static: true" this time.

This commit ensures that the `static-query` migration schematic
now handles this cases properly. Also template property reads
that access queries from within a `<ng-template>` are ignored
as these can't access the query before the view has been initialized.

**Note**: Blocked on #29688 

Resolves FW-1216

devversion committed:
refactor(core): move schematic component template visitor to utils

kara commented on packages/core/schematics/test/static_queries_migration_spec.ts:
```diff
@@ -1217,6 +1217,140 @@ describe('static-queries migration', () => {
+
+         writeFile(`/my-template.html`, `
+        <foo #test></foo>
+        <comp [dir]="myCtx.query"></comp>
```
I'm a little confused about this test case. What is `myCtx` here? Seems like it would be `undefined`?

devversion replied:
It's not relevant if `myCtx` exists or not for this test-case. Just added it to make the test-case more clear. Also added a small comment.

The reason this test-case exists is that the `render3.PropertyRead` AST objects for the following expressions: `this.query` and `myProp.query` are very similar and the query detection could accidentally consider `myProp.query` as reference to the actual query. This is something that I ran into and therefore added a test-case. The `receiver` of the `render3.PropertyRead` needs to be `this` (aka `implicit`) and this test-case ensures that we properly handle the inline expressions.

See [here](https://github.com/angular/angular/pull/29713/files#diff-88e0de5c27ce50296eee25472fee45a9R81)

kara reviewed (approved):
LGTM

devversion committed:
refactor(core): static-query schematic should check templates

Queries can technically be also accessed within component templates
e.g.

```html
<my-comp [binding]="myQuery"></my-comp>
```

In that case the query with the property "myQuery" is accessed
statically and needs to be marked with `static: true`. There are
other edge cases that need to be handled as the template property
read doesn't necessarily resolve to the actual query property.

For example:

```html
<foo #myQuery></foo>
<my-comp [binding]="myQuery"></my-comp>
```

In this scenario the binding doesn't refer to the actual query
because the template reference variable takes precedence. The
query doesn't need to be marked with "static: true" this time.

This commit ensures that the `static-query` migration schematic
now handles this cases properly. Also template property reads
that access queries from within a `<ng-template>` are ignored
as these can't access the query before the view has been initialized.

Resolves FW-1216

kara commented:
[presubmit](http://test/OCL:242573947:BASE:242574017:1554770363513:9fe1fb05)

kara commented:
@devversion This one needs a rebase too

devversion commented:
Waiting with this one as it needs another rebase after #29688 has been merged.

devversion commented:
@kara Rebased. Should be ready now.

alxhub commented on packages/core/schematics/migrations/static-queries/angular/query_read_html_visitor.ts:
```diff
@@ -0,0 +1,88 @@
+  visitPropertyRead(node: PropertyRead, span: ParseSourceSpan): any {
+    // The receiver of the property read needs to be "implicit" as queries are accessed
+    // from the component instance and not from other objects.
+    if (node.receiver instanceof ImplicitReceiver && node.name === this.queryPropertyName) {
```
FYI, `ImplicitReceiver` is usually the component instance, but doesn't have to be.

```html
<div *ngFor="let foo of foos">
{{foo}}
</div>
```

Here, the `{{foo}}` binding is a `PropertyRead` from the `ImplicitReceiver`, even though it's reading from a variable and not from the component instance.

devversion replied:
Thanks for the heads up! I actually just double-checked and it looks like we are good here because we only visit property read nodes that aren't part of templates. So there can't be any template input variable that falsifies this check.
